File: 20_valid_parentheses/src.c
#include <stdio.h>
#include <string.h>

int isValid(char* s) {
    char stack[strlen(s)];
    int head = -1;
    for(int i = 0; i < strlen(s); i++) {
        if (head == -1) {
            head++;
            stack[head] = s[i];
            continue;
        }
        if (s[i] == ')' && stack[head] == '(') {
            head--;
        }
        else if (s[i] == '}' && stack[head] == '{') {
            head--;
        }
        else if (s[i] == ']' && stack[head] == '[') {
            head--;
        }
        else {
            head++;
            stack[head] = s[i];
        }
    }    
    if (head == -1) {
        return 1;
    } else {
        return 0;
    }
    return 0;
}

int main() {
    // char *tmp = "()";
    // printf("%d\n", isValid(tmp));
    // char *tmp = "()[]{}";
    // printf("%d\n", isValid(tmp));
    // char *tmp = "(]";
    // printf("%d\n", isValid(tmp));
    // char *tmp = "(])";
    // printf("%d\n", isValid(tmp));
    char *tmp = "{[]}";
    printf("%d\n", isValid(tmp));
}

File: 20_valid_parentheses/src.py
class Solution:
    def isValid(self, s: str) -> bool:
        data = {")": "(", "}": "{", "]": "["}
        stack = []
        for i in s:
            if not stack:
                stack.append(i)
                continue
            if i in data.keys():
                if stack[-1] == data[i]:
                    stack.pop()
                    continue
                else:
                    stack.append(i)
            else:
                stack.append(i)
        return not bool(stack)


solution = Solution()
print(solution.isValid("()"))
print(solution.isValid("()[]{}"))
print(solution.isValid("(]"))
print(solution.isValid("(])"))
print(solution.isValid("{[]}"))
